pkg/model/deployment_chain.go

```diff
@@ -29,3 +31,15 @@ func (dc *DeploymentChain) IsSuccessfullyCompletedBlock(blockIndex uint32) (bool
 	}
 	return true, nil
 }
+
+func (dc *DeploymentChain) ListIncompletedNodes() []*ChainNode {
+	icn := make([]*ChainNode, 0)
+	for _, block := range dc.Blocks {
+		for _, node := range block.Nodes {
+			if !IsCompletedDeployment(node.DeploymentRef.Status) {
```

Comment:
Here seems to be on the premise that DeploymentRef isn't nil, but ChainNode.ChainDeploymentRef isn't guaranteed to be required.

How about adding validation to that field in deployment_chain.proto?

Reply:
Ah, good catch. In fact, we should add a kind of `if node.DeploymentRef == nil` check here and count it as uncompleted deployment since the deploymentRef is created by the pipedx (who is in charge of this application) so we can not require it as not nil in the deployment chain model at the time the deployment chain had just created.

Reply:
On second thought, I just realized it's a bit complicated here. In case the deployment ref has not yet been created, we can not add it to the IncompletedNodes slide, since we need those deploymentRef fields to send CANCEL_DEPLOYMENT to those pipedx

Reply:
I see. In that case, I feel we need to add it at L576 in `pkg/app/api/grpcapi/piped_api.go` as well.

Reply:
Alright. Got it 👍

Reply:
I have another idea, instead of creating CANCEL_DEPLOYMENT for all applications here, we can delay that CANCEL_DEPLOYMENT command creation until pipedx asks control-plane whether it should start planning the deployment (via InChainDeploymentPlannable). In that case, we can ensure that the deployment exists thus deploymentRef is available in the deployment chain model. How do you think about that?

Reply:
Addressed by https://github.com/pipe-cd/pipe/pull/2866/commits/1a55040949aeaeecd8bbd7066f6844d5670a5303 , PTAL when you have time 😉